### PR TITLE
[SPARK-8218][SQL] Add binary log math function

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-8218

Because there is already `log` unary function defined, the binary log function is called `logarithm` for now.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1299,6 +1299,70 @@ object functions {
    */
   def toRadians(columnName: String): Column = toRadians(Column(columnName))
 
+  /**
+   * Returns the first argument-base logarithm of the second argument.
+   *
+   * @group math_funcs
+   * @since 1.4.0
+   */
+  def logarithm(l: Column, r: Column): Column = Logarithm(l.expr, r.expr)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: let's name the 1st argument as "base", and the second as "a".

and also only keep the following 2 versions:

log(base: Double, a: Column)

log(base: Double, a: String)

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: btw I think it is OK to just name this log, and we should move them to where log is right now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -122,6 +122,7 @@ object FunctionRegistry {
     expression[Tanh]("tanh"),
     expression[ToDegrees]("todegrees"),
     expression[ToRadians]("toradians"),
+    expression[Logarithm]("logarithm"),
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: we should name this log to make it compatible. (you'd need to make the logarithm expression itself being able to handle both)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 3]
```diff
@@ -202,3 +202,27 @@ case class Pow(left: Expression, right: Expression)
       """
   }
 }
+
+object Logarithm {
+  def apply(exprs: Seq[Expression]): Expression = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: we don't need this one do we? just need an apply with one argument, along with the original case class' constructor.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 4]
```diff
@@ -202,3 +202,27 @@ case class Pow(left: Expression, right: Expression)
...
+      sys.error(s"Log only accepts two input expressions")
+    }
+  }
+  def apply(child: Expression): Expression = new Log(child)
+}
+
+case class Logarithm(left: Expression, right: Expression)
+  extends BinaryMathExpression((c1, c2) => math.log(c2) / math.log(c1), "LOG") {
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    defineCodeGen(ctx, ev, (c1, c2) => s"java.lang.Math.log($c2) / java.lang.Math.log($c1)") + s"""
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: let's generate two versions of the code, i.e. if there is only one argument, generate the old version directly. When there are two arguments, generate the new version.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 5]
```diff
@@ -298,6 +298,17 @@ class SQLQuerySuite extends QueryTest with BeforeAndAfterAll with SQLTestUtils {
     )
   }
 
+  test("LOG") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: let's remove this test case, and use selectExpr in DataFrameFunctionSuite to test the parser.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 4 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: do you mean, when there is only one argument, `Logarithm` generates the codes as the same as `Log`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: yes, exactly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 6]
```diff
@@ -254,3 +254,54 @@ case class Pow(left: Expression, right: Expression)
       """
   }
 }
+
+object Logarithm {
+  def apply(child: Expression): Expression = new Log(child)
+}
+
+case class Logarithm(left: Expression, right: Expression)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: We seem to be doing this throughout the file, but it seems pretty confusing to me to be using `left` and `right` instead of something like `value` and `base`.  I'm not sure this is worth whatever code reuse we are getting.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 7]
```diff
@@ -254,3 +254,54 @@ case class Pow(left: Expression, right: Expression)
...
+  extends BinaryMathExpression((c1, c2) => math.log(c2) / math.log(c1), "LOG") {
+  override def eval(input: Row): Any = {
+    val evalE2 = right.eval(input)
+    if (evalE2 == null) {
+      null
+    } else {
+      val evalE1 = left.eval(input)
+      var result: Double = 0.0
+      if (evalE1 == null) {
+        result = math.log(evalE2.asInstanceOf[Double])
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Is it standard to assume base 10 if the exponent is null?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 6 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: +1

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Oh - one thing is that left/right is coming from BinaryExpression

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 8]
```diff
@@ -110,6 +110,19 @@ class DataFrameFunctionsSuite extends QueryTest {
       testData2.collect().toSeq.map(r => Row(~r.getInt(0))))
   }
 
+  test("log") {
+    val df = Seq[(Integer, Integer)]((123, null)).toDF("a", "b")
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: can you move this test into the math expression suite?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 9]
```diff
@@ -254,3 +254,54 @@ case class Pow(left: Expression, right: Expression)
...
+      } else {
+        result = math.log(evalE2.asInstanceOf[Double]) / math.log(evalE1.asInstanceOf[Double])
+      }
+      if (result.isNaN) null else result
+    }
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    if (left.dataType != right.dataType) {
+      // log.warn(s"${left.dataType} != ${right.dataType}")
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Remove

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 10]
```diff
@@ -110,6 +110,19 @@ class DataFrameFunctionsSuite extends QueryTest {
...
+    checkAnswer(
+      df.select(org.apache.spark.sql.functions.log("a"),
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Nit: When wrapping, wrap all arguments.  Also the indent below is wrong.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 11]
```diff
@@ -204,4 +204,18 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     testBinary(Atan2, math.atan2)
   }
 
+  test("binary log") {
+    val f = (c1: Double, c2: Double) => math.log(c2) / math.log(c1)
+    val domain = (1 to 20).map(v => (v * 0.1, v * 0.2))
+
+    domain.foreach { case (v1, v2) =>
+        checkEvaluation(Logarithm(Literal(v1), Literal(v2)), f(v1 + 0.0, v2 + 0.0), EmptyRow)
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Indent is wrong.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 6 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Yeah, that is what I meant saying it wasn't worth whatever code reuse we are getting.  The other option would be to name the arguments and have `def left = base`, etc.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 12]
```diff
@@ -204,4 +204,18 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+        checkEvaluation(Logarithm(Literal(v2), Literal(v1)), f(v2 + 0.0, v1 + 0.0), EmptyRow)
+    }
+    // When base is null, Logarithm is as same as Log
+    checkEvaluation(Logarithm(Literal.create(null, DoubleType), Literal(1.0)),
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Same comment about wrapping.

**Comment by viirya** [comment]
@rxin @marmbrus I think your comments are addressed now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 13]
```diff
@@ -254,3 +254,52 @@ case class Pow(left: Expression, right: Expression)
       """
   }
 }
+
+object Logarithm {
+  def apply(child: Expression): Expression = new Log(child)
+}
+
+case class Logarithm(left: Expression, right: Expression)
+  extends BinaryMathExpression((c1, c2) => math.log(c2) / math.log(c1), "LOG") {
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: If we need to override the `eval()`, probably it's better to inherit from `BinaryArithmetic`, not `BinaryMathExpression`, then we needn't to pass the additional function object for its parent class.
Or can we remove the `eval()` from this class?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 14]
```diff
@@ -1084,6 +1084,22 @@ object functions {
...
+  def log(base: Double, a: Column): Column = Logarithm(lit(base).expr, a.expr)
+
+  /**
+   * Returns the first argument-base logarithm of the second argument.
+   *
+   * @group math_funcs
+   * @since 1.4.0
+   */
+  def log(base: Double, a: String): Column = log(base, Column(a))
+
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Why don't we support specify the `base` from a column?

File: sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala [hunk 15]
```diff
@@ -236,6 +236,19 @@ class MathExpressionsSuite extends QueryTest {
     testOneToOneNonNegativeMathFunction(log1p, math.log1p)
   }
 
+  test("binary log") {
+    val df = Seq[(Integer, Integer)]((123, null)).toDF("a", "b")
+    checkAnswer(
+      df.select(org.apache.spark.sql.functions.log("a"),
```
Comment (chenghao-intel) [inline sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala]: Can we remove the `org.apache.spark.sql.functions.` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 16]
```diff
@@ -254,3 +254,52 @@ case class Pow(left: Expression, right: Expression)
       """
   }
 }
+
+object Logarithm {
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Do we really need this? We should assume people will use the `ln` instead?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 17]
```diff
@@ -204,4 +204,22 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
 
+  test("binary log") {
+    val f = (c1: Double, c2: Double) => math.log(c2) / math.log(c1)
+    val domain = (1 to 20).map(v => (v * 0.1, v * 0.2))
+
+    domain.foreach { case (v1, v2) =>
+      checkEvaluation(Logarithm(Literal(v1), Literal(v2)), f(v1 + 0.0, v2 + 0.0), EmptyRow)
+      checkEvaluation(Logarithm(Literal(v2), Literal(v1)), f(v2 + 0.0, v1 + 0.0), EmptyRow)
+    }
+    // When base is null, Logarithm is 10-based log.
```
Comment (chenghao-intel) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Why not `null` value if the base is `null`?

File: sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala (hunk 15 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala]: It conflicts with Logger.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala (hunk 17 above)
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: We support the default base as 10.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 14 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: It is suggested by @rxin. I think it is reasonable because it is hard to have a use case to returns the logarithm of one column with another column as base. Usually you want to compute the logarithm values for a column with the same base.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 16 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Because we want to support the usage of `df.selectExpr("log(a)", "log(2.0, a)", "log(b)")`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 13 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: I think `BinaryArithmetic` is more limited in its semantic meaning. As I can tell, it is more suitable for the binary expression between two expressions of same type. `BinaryMathExpression` is more like to represent math function with two expressions.

Due to support the case when the left is null in that 10 base logarithm is applied, to override `eval()` is needed here.

**Comment by viirya** [comment]
@rxin is this ready to go?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 13 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Okay, but this is confusing.  Is it this lambda that is being used for evaluation or the `eval` function below?  We should not have both if only one is used.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala (hunk 17 above)
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Using null as a sentinel value to signify "use the default" does not match any of the other code in Spark SQL.  Typically all operations that involve null return null.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 6 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Inheriting from `BinaryMathExpression` seems like a bad idea to me.  It is forcing the arguments to have weird names and is resulting in dead code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 13 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: It should be ok because there is other binary math expression `Atan2`  that overrides `eval` because its eval behavior is different than the default one in `BinaryMathExpression`.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: I would argue that that function is also implemented in a confusing way.  We should not shoehorn things into the class hierarchy if its going to result hard to follow code.  I'd rather we have small amounts of code duplication.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala (hunk 17 above)
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: If so, we should not support the default 10 based logarithm? Since it is suggested by @rxin, how do you think?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 13 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: OK. Sounds reasonable. I think I can refactor this part of codes a little.

**Comment by rxin** [comment]
log(a) should be log(e, a), where e is the natural logatirhm. so we should just have a ctor that takes one argument, and creates Log(EulerNumber, argument).

**Comment by viirya** [comment]
@rxin it is. Currently log(a) will create Log(a), which is natural logarithm. The issue now is how to handle `log(null, a)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 16 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: take a look at https://github.com/apache/spark/pull/6806

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: ok. it solves this problem.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 18]
```diff
@@ -255,3 +265,56 @@ case class Pow(left: Expression, right: Expression)
       """
   }
 }
+
+case class Logarithm(left: Expression, right: Expression)
+  extends AbstractBinaryMathExpression[Double, Double, Double]("LOG") {
+  def this(child: Expression) = {
+    this(Literal(math.E), child)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: don't use a literla, just use EulerNumber expression

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala (hunk 17 above)
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Maybe I misunderstand @rxin's suggestion. I think you were suggesting the ctor that I just committed, not a default n base logarithm for log(null, a), right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 18 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: ok. haven't  noticed there is EulerNumber.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala (hunk 17 above)
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: if the 1st argument is null, we definitely shouldn't use 10 base implicitly. can you take a look at what hive returns in that case? maybe just null.

Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: Yes. it is null. I will update this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 19]
```diff
@@ -90,18 +90,26 @@ abstract class UnaryMathExpression(f: Double => Double, name: String)
 }
 
 /**
+ * A binary expression specifically for math functions that take two input parameters and
+ * returns one output value.
+ * @param name The short name of the function
+ */
+abstract class AbstractBinaryMathExpression[T, U, V](name: String)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: @rxin, do you think I need to revert this? It is now just used by `Atan2`.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: yes let's not complicate the type hierarchy just for one thing

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 20]
```diff
@@ -1084,6 +1084,22 @@ object functions {
...
+   */
+  def log(base: Double, a: Column): Column = Logarithm(lit(base).expr, a.expr)
+
+  /**
+   * Returns the first argument-base logarithm of the second argument.
+   *
+   * @group math_funcs
+   * @since 1.4.0
+   */
+  def log(base: Double, a: String): Column = log(base, Column(a))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: can we rename a to columnName here

File: python/pyspark/sql/functions.py [hunk 21]
```diff
@@ -143,7 +143,8 @@ def _():
     'atan2': 'Returns the angle theta from the conversion of rectangular coordinates (x, y) to' +
              'polar coordinates (r, theta).',
     'hypot': 'Computes `sqrt(a^2^ + b^2^)` without intermediate overflow or underflow.',
-    'pow': 'Returns the value of the first argument raised to the power of the second argument.'
+    'pow': 'Returns the value of the first argument raised to the power of the second argument.',
+    'log': 'Returns the first argument-based logarithm of the second argument',
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: can we define this function directly instead of using this magic?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 22]
```diff
@@ -255,3 +255,18 @@ case class Pow(left: Expression, right: Expression)
...
 }
+
+case class Logarithm(left: Expression, right: Expression)
+  extends BinaryMathExpression((c1, c2) => math.log(c2) / math.log(c1), "LOG") {
+  def this(child: Expression) = {
+    this(EulerNumber(), child)
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    defineCodeGen(ctx, ev, (c1, c2) => s"java.lang.Math.log($c2) / java.lang.Math.log($c1)") + s"""
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: if left == EulerNumber, can we generate the code to just use java.lang.Math.log?

File: python/pyspark/sql/functions.py [hunk 23]
```diff
@@ -404,6 +405,21 @@ def when(condition, value):
 
 
 @since(1.4)
```
Comment (davies) [inline python/pyspark/sql/functions.py]: 1.5

**Comment by rxin** [comment]
Thanks - I'm going to merge this and fix the since version.

**Comment by viirya** [comment]
Thanks.
